feat(inventory) change report

jeremielodi opened the pull request:
closes #3975

jeremielodi committed:
feat(inventory) change report

jeremielodi committed:
closes #3975

jniles commented on client/src/modules/reports/generate/inventoryChange/inventoryChange.config.js:
```diff
@@ -0,0 +1,76 @@
+  const reportUrl = 'reports/inventory/changes/';
+
+  vm.previewGenerated = false;
+  vm.reportDetails = {
```
These looks like they are copied from somewhere.  Please remove all the code you didn't write and didn't use.

jniles commented on client/src/i18n/en/report.json:
```diff
@@ -106,6 +106,10 @@
       "DESCRIPTION": "This report details the amount to paid in every inventory group for an invoice during a period, for a specific debor group"
     },
+    "INVENTORY_CHANGE" : {
+      "TITLE" : "Inventory change report",
```
Let's call this "Inventory Changes Report"

jniles commented on client/src/modules/reports/generate/inventoryChange/inventoryChange.config.js:
```diff
@@ -0,0 +1,76 @@
+    vm.reportDetails.shouldShowPaidDetails = value;
+  };
+
+  vm.onChangeShowInvoicedDetails = value => {
```
It doesn't look like you use any of this.  Please remove anything you don't use or need.

jniles commented on server/controllers/inventory/reports/changes.js:
```diff
@@ -0,0 +1,130 @@
+
+/**
+ * @overview Price Report
```
I don't think this is a price report.  Please update or remove these comments.

jniles commented on server/models/migrations/next/migrate.sql:
```diff
@@ -191,3 +191,13 @@ INSERT INTO `survey_form_type` (`id`, `label`, `type`, `is_list`) VALUES
+--- by jeremielodi
+--- 2019-11-18
+
+INSERT INTO unit VALUES
```
👍

jniles commented on server/controllers/inventory/reports/changes.js:
```diff
@@ -0,0 +1,130 @@
+  });
+  const metadata = _.clone(req.session);
+
+  let report;
```
You can move this `let` into a `const` in the `try/catch` block.

jniles commented on server/controllers/inventory/reports/changes.js:
```diff
@@ -0,0 +1,130 @@
+    const { dateFrom, dateTo } = params;
+
+    const inventorySql = `
+      SELECT DISTINCT BUID(iv.uuid) as uuid, iv.text AS inventory_name
```
Why do we need the current inventory items?  Can't we just use the data in the logs?

jniles commented on server/controllers/inventory/reports/changes.js:
```diff
@@ -0,0 +1,130 @@
+
+function formatKeys(record) {
+  const removables = ['group_uuid', 'type_id', 'unit_id'];
+  removables.forEach(r => {
```
Just use lodash:
```js
_.omit(record, removables);
```

jniles commented on server/controllers/inventory/reports/changes.js:
```diff
@@ -0,0 +1,130 @@
+    delete record[r];
+  });
+  if (record.text) {
+    record.label = _.clone(record.text);
```
I don't think you need `_.clone` here.  I think you can just set them equivalent.

jniles commented on server/controllers/inventory/reports/changes.js:
```diff
@@ -0,0 +1,130 @@
+    record.label = _.clone(record.text);
+    delete record.text;
+  }
+  return record;
```
It looks like your entire `formatKeys` function could be:
```js
function formatKeys(record) {
  record.label = record.text;
  return _.omit(record, ['group_uuid', 'type_id', 'unit_id', 'text']);
}
```

Try it and see if that works.

jniles reviewed (changes requested):
Looks like there is quite a few lines copied from other reports that make this PR hard to read and evaluate.  Could you clean those up and ping me back for a review?

jniles committed:
fix(reports): address review feedback

This commit addresses review feedback to make sure the PR passes muster.

jniles committed:
fix(inventory): rename of inventoryColMap

The inventory changes report might have never worked, since it was
referencing an object in the inventory/core that doesn't exist.  We've
changed the reference to the correct value.

jniles committed:
Merge remote-tracking branch 'upstream/master' into inventory-change-report
